File: java/704-binary-search.java
class Solution {
    public int search(int[] nums, int target) {

        if(nums.length==1 && nums[0] == target){
            return 0;
        }

        int l = 0;
        int r = nums.length-1;
        while(l<=r){
            int mid = (l+r)/2;
            if(nums[mid] == target){
                return mid;
            }

            if(nums[mid]>target){
                r = mid-1;
            }else{
                l = mid+1;
            }
        }   
        return -1;
    }
}

File: kotlin/704-Binary-Search.kt
class Solution {
    // Easy solution
    /*
    fun search(nums: IntArray, target: Int): Int {
        return nums.indexOf(target)
    }
    */
    
    // Recursive solution
    fun search(nums: IntArray, target: Int): Int {
        return searchAux(nums, target, 0, nums.size - 1)
    }

    private fun searchAux(nums: IntArray, target: Int, begin: Int, end: Int): Int {
        if (begin > end) return -1
        
        val mid: Int = (begin + end) / 2
        
        return if (nums[mid] == target) mid
        else if (nums[mid] > target) searchAux(nums, target, begin, mid - 1)
        else searchAux(nums, target, mid + 1, end)
    }
}
